File: internal/observability/byteopacity.go
package observability

import (
	"context"
	"log/slog"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
)

// BumpCodecRecordDecode is the canonical entry point for the
// byte-opacity tripwire. Every code path that decodes an individual
// record (rather than treating the RecordBatch as opaque bytes) MUST
// call this — and that should be never. site is a short string that
// names the offender so the alert payload is actionable.
//
// As of v1, no skafka code path calls this. If you find yourself
// adding the first call, stop: the v3.3 plan is explicit that the
// broker is a byte mover, not a byte interpreter. A use case that
// genuinely requires record-level inspection (e.g. a future ksqlDB
// integration) needs a separate design discussion before it lands.
func BumpCodecRecordDecode(ctx context.Context, site string) {
	Global().CodecRecordDecode.Add(ctx, 1,
		metric.WithAttributes(attribute.String("site", site)))
	slog.Warn("byte-opacity tripwire fired (record decode)", "site", site)
}

// BumpCodecBatchReencode is the sibling tripwire for re-encoding a
// RecordBatch. Same contract as BumpCodecRecordDecode: every increment
// is a bug, and the slog.Warn line is meant to surface in production
// logs so an investigation starts before alerts even fire.
func BumpCodecBatchReencode(ctx context.Context, site string) {
	Global().CodecBatchReencode.Add(ctx, 1,
		metric.WithAttributes(attribute.String("site", site)))
	slog.Warn("byte-opacity tripwire fired (batch reencode)", "site", site)
}

File: crates/sk-observability/src/byteopacity.rs
//! Byte-opacity tripwire counters.
//!
//! Port of `archive/internal/observability/byteopacity.go`. The
//! rewrite plan's load-bearing invariant is "the broker is a byte
//! mover, not a byte interpreter": no code path should decode
//! individual records or re-encode a `RecordBatch`. These counters
//! MUST stay at zero in steady state — every increment is a bug and
//! the `SkafkaByteOpacityViolated` alert fires on non-zero.
//!
//! As of v1 no skafka code path calls these. If you find yourself
//! adding the first call, stop: a use case that genuinely requires
//! record-level inspection needs a separate design discussion before
//! it lands. See [`crate::metrics::Metrics::codec_record_decode`].

use opentelemetry::KeyValue;

/// Record a byte-opacity tripwire (record decode). `site` names the
/// offending code path so the alert payload is actionable.
pub fn bump_codec_record_decode(site: &'static str) {
    let m = crate::metrics::global();
    m.codec_record_decode.add(1, &[KeyValue::new("site", site)]);
    tracing::warn!(site, "byte-opacity tripwire fired (record decode)");
}

/// Record a byte-opacity tripwire (batch re-encode). `site` names the
/// offending code path so the alert payload is actionable.
pub fn bump_codec_batch_reencode(site: &'static str) {
    let m = crate::metrics::global();
    m.codec_batch_reencode
        .add(1, &[KeyValue::new("site", site)]);
    tracing::warn!(site, "byte-opacity tripwire fired (batch reencode)");
}
